### PR TITLE
feat: add `GOOGLE_CLOUD_CPP_ENABLE` CMake variable

### PR DESCRIPTION
**Pull request by remyabel**
Fixes: #5511

cc: @devjgm

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517)

### CODE REVIEW

**Comment by coryan** [comment]
This is getting complicated enough that maybe we should have tests for it.  Admittedly, the same could be said for other CMake snippets.

No action required, the intent of this comment is to get @devjgm and myself thinking about how we test CMake stuff.

File: CMakeLists.txt
Comment (devjgm) [inline CMakeLists.txt:158]: Note: Here's the right place for us to add a warning message that these per-library enable flags should be avoided in favor of `GOOGLE_CLOUD_CPP_ENABLE`.  I think we may want to add a message for that.

@coryan WDYT?

File: CMakeLists.txt
Comment (devjgm) [inline CMakeLists.txt:148]: My CMake skills are intermediate at best, so I'll ask: Is there a reason to have `GOOGLE_CLOUD_CPP_LIBRARIES` in addition to `GOOGLE_CLOUD_CPP_ENABLE`? Would it work have only the latter variable?

**Review by devjgm — COMMENTED** [review]
@remyabel I know this is a draft, but I'll say I think this looks really good and I think it'll be a better UX for our customers. @coryan is much better at CMake than I am, so I'd like to get his input also before we merge. Thank you!!

File: CMakeLists.txt
Comment (remyabel) [reply CMakeLists.txt]: I don't remember. Fixed.

File: CMakeLists.txt
Comment (remyabel) [reply CMakeLists.txt:160]: Fixed.

**Comment by remyabel** [comment]
> I know this is a draft

There is still one minor problem. If one of the variables is overridden by the user, then `GOOGLE_CLOUD_CPP_ENABLE` doesn't take precedence over the cache. Example (imagine a series of invocations):

```cmake
-DGOOGLE_CLOUD_CPP_ENABLE=storage,spanner,pubsub -DGOOGLE_CLOUD_CPP_ENABLE_STORAGE=OFF
# spanner,pubsub
-DGOOGLE_CLOUD_CPP_ENABLE=storage,spanner,pubsub
# spanner,pubsub
-DGOOGLE_CLOUD_CPP_ENABLE=storage,spanner,pubsub -DGOOGLE_CLOUD_CPP_ENABLE_STORAGE=ON
# storage,spanner,pubsub
-DGOOGLE_CLOUD_CPP_ENABLE=spanner,pubsub
# spanner,pubsub
-DGOOGLE_CLOUD_CPP_ENABLE=storage,spanner,pubsub
# storage,spanner,pubsub
```

Although this might not be an issue in practice due to the warning.

**Review by coryan — CHANGES_REQUESTED** [review]
I think we can ignore the multiple invocations with inconsistent settings in CMake.  

There are some additional places where we use the `*_ENABLE_${library}` flags, maybe do a search for all uses everywhere?

> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517#-:-MMWrv6r8NLAlq_AMCDE:bfm1h6a)* status: all files reviewed, 5 unresolved discussions (waiting on @devjgm and @remyabel)

___
*[CMakeLists.txt, line 158 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517#-MMPXqiAKTdy2kkbHe1F-r1-158:-MMWsXFmEfA7myTGdGKw:b-85i64v) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/4c806b3d7a5b9955290b5fec50f0fbddac7cc95d/CMakeLists.txt#L158)):*
<details><summary><i>Previously, remyabel wrote…</i></summary><blockquote>

Fixed.
</blockquote></details>

LGTM
___
*[CMakeLists.txt, line 157 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517#-MMWsKpm9yEUSmr0auiU:-MMWsKpm9yEUSmr0auiV:b-vihcvx) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0b16fdaffb5a4c00d1fbeadc6f3140647e239cc8/CMakeLists.txt#L157)):*
> ```CMake
>         message(
>             WARNING
>                 "Using ${_library} per-library enable flag is discouraged in "
> ```

I think just saying "Using GOOGLE_CLOUD_CPP_ENABLE_${_library} ..." is clearer.
___
*[CMakeLists.txt, line 169 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517#-MMWtDjgAW1QNWYIuCba:-MMWtDjgAW1QNWYIuCbb:bj4odr4) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0b16fdaffb5a4c00d1fbeadc6f3140647e239cc8/CMakeLists.txt#L169)):*
> ```CMake
> string(
>     CONCAT GOOGLE_CLOUD_CPP_ENABLE_GRPC_EXPRESSION
>            "GOOGLE_CLOUD_CPP_ENABLE_BIGQUERY OR "
> ```

I think this becomes:

```cmake
string(CONCAT GOOGLE_CLOUD_CPP_ENABLE_GRPC_EXPRESSION
    "(bigquery IN_LIST GOOGLE_CLOUD_CPP_ENABLE) OR "
    "(spanner IN_LIST GOOGLE_CLOUD_CPP_ENABLE) OR "
    ... and so on ...
```
___
*[CMakeLists.txt, line 195 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517#-MMWsdhj2S0hplLFixcm:-MMWsdhj2S0hplLFixcn:b-q4mg0l) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0b16fdaffb5a4c00d1fbeadc6f3140647e239cc8/CMakeLists.txt#L195)):*
> ```CMake
> 
> add_subdirectory(google/cloud)
> foreach (library ${GOOGLE_CLOUD_CPP_ENABLE})
> ```

FYI, I did not think of a loop here, great idea, looks cool.
___
*[CMakeLists.txt, line 206 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517#-MMWtwAi87lQgieJghwf:-MMWtwAi87lQgieJghwg:bgqin0y) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0b16fdaffb5a4c00d1fbeadc6f3140647e239cc8/CMakeLists.txt#L206)):*
> ```CMake
> # to tradeoff readability vs. "making them compile everywhere".
> if (GOOGLE_CLOUD_CPP_ENABLE_CXX_EXCEPTIONS
>     AND GOOGLE_CLOUD_CPP_ENABLE_BIGTABLE
> ```

this needs to become:

```cmake
if (GOOGLE_CLOUD_CPP_ENABLE_CXX_EXCEPTIONS
    AND bigtable IN_LIST GOOGLE_CLOUD_CPP_ENABLE
    AND storage IN_LINE GOOGLE_CLOUD_CPP_ENABLE)
```

**Review by remyabel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517)* status: 0 of 1 files reviewed, 5 unresolved discussions (waiting on @coryan and @devjgm)

___
*[CMakeLists.txt, line 157 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517#-MMWsKpm9yEUSmr0auiU:-MMY4ReX-E_WQGe4N48M:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0b16fdaffb5a4c00d1fbeadc6f3140647e239cc8/CMakeLists.txt#L157)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I think just saying "Using GOOGLE_CLOUD_CPP_ENABLE_${_library} ..." is clearer.
</blockquote></details>

Done.
___
*[CMakeLists.txt, line 169 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517#-MMWtDjgAW1QNWYIuCba:-MMY4VuTDaiyFjSUzKmq:bl0hh4g) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0b16fdaffb5a4c00d1fbeadc6f3140647e239cc8/CMakeLists.txt#L169)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I think this becomes:

```cmake
string(CONCAT GOOGLE_CLOUD_CPP_ENABLE_GRPC_EXPRESSION
    "(bigquery IN_LIST GOOGLE_CLOUD_CPP_ENABLE) OR "
    "(spanner IN_LIST GOOGLE_CLOUD_CPP_ENABLE) OR "
    ... and so on ...
```


</blockquote></details>

I do not know why this doesn't work. I uploaded the debugging code, maybe you can spot something I missed.
___
*[CMakeLists.txt, line 206 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517#-MMWtwAi87lQgieJghwf:-MMY4V3ZCZyIaM0k6w6B:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0b16fdaffb5a4c00d1fbeadc6f3140647e239cc8/CMakeLists.txt#L206)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

this needs to become:

```cmake
if (GOOGLE_CLOUD_CPP_ENABLE_CXX_EXCEPTIONS
    AND bigtable IN_LIST GOOGLE_CLOUD_CPP_ENABLE
    AND storage IN_LINE GOOGLE_CLOUD_CPP_ENABLE)
```
</blockquote></details>

Done.

**Review by remyabel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517)* status: 0 of 1 files reviewed, 5 unresolved discussions (waiting on @coryan and @devjgm)

___
*[CMakeLists.txt, line 169 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517#-MMWtDjgAW1QNWYIuCba:-MMY5SjkEcOZT7Hb6UZ6:b-6rt4ta) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0b16fdaffb5a4c00d1fbeadc6f3140647e239cc8/CMakeLists.txt#L169)):*
<details><summary><i>Previously, remyabel wrote…</i></summary><blockquote>

I do not know why this doesn't work. I uploaded the debugging code, maybe you can spot something I missed.
</blockquote></details>

Sorry, I should've been more specific. `${GOOGLE_CLOUD_CPP_ENABLE_GRPC_EXPRESSION}` seems to evaluate to `OFF` (first if expression) while the bare expression seems to evaluate to `ON` (second if expression). Perhaps I made a typo but I can't spot it.

**Review by coryan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517)* status: 0 of 1 files reviewed, 4 unresolved discussions (waiting on @coryan, @devjgm, and @remyabel)

___
*[CMakeLists.txt, line 169 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517#-MMWtDjgAW1QNWYIuCba:-MMY6gTJBlrphq5oq8g_:bspzw39) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0b16fdaffb5a4c00d1fbeadc6f3140647e239cc8/CMakeLists.txt#L169)):*
<details><summary><i>Previously, remyabel wrote…</i></summary><blockquote>

Sorry, I should've been more specific. `${GOOGLE_CLOUD_CPP_ENABLE_GRPC_EXPRESSION}` seems to evaluate to `OFF` (first if expression) while the bare expression seems to evaluate to `ON` (second if expression). Perhaps I made a typo but I can't spot it.
</blockquote></details>

Maybe the weird rules around `if()` in CMake:

https://cmake.org/cmake/help/latest/command/if.html#variable-expansion

It might be simpler to create a helper function, something like:

```cmake
function(google_cloud_cpp_set_enable_grpc_default)
  set(value OFF)
  for (_library IN bigquery bigtable spanner pubsub generator)
    if (bigquery IN_LIST GOOGLE_CLOUD_CPP_ENABLE)
      set(value ON)
    endif ()
  endfor ()
  set(GOOGLE_CLOUD_CPP_ENABLE_GRPC_DEFAULT "${value}" PARENT_SCOPE)
endfunction()

google_cloud_cpp_set_enable_grpc_default()
```

it may work better *and* it seems easier to understand / debug. WDYT?

**Review by remyabel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517)* status: 0 of 1 files reviewed, 4 unresolved discussions (waiting on @coryan and @devjgm)

___
*[CMakeLists.txt, line 169 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5517#-MMWtDjgAW1QNWYIuCba:-MMYtZzg-23PdxRh3AMN:bz9rjwu) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/0b16fdaffb5a4c00d1fbeadc6f3140647e239cc8/CMakeLists.txt#L169)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Maybe the weird rules around `if()` in CMake:

https://cmake.org/cmake/help/latest/command/if.html#variable-expansion

It might be simpler to create a helper function, something like:

```cmake
function(google_cloud_cpp_set_enable_grpc_default)
  set(value OFF)
  for (_library IN bigquery bigtable spanner pubsub generator)
    if (bigquery IN_LIST GOOGLE_CLOUD_CPP_ENABLE)
      set(value ON)
    endif ()
  endfor ()
  set(GOOGLE_CLOUD_CPP_ENABLE_GRPC_DEFAULT "${value}" PARENT_SCOPE)
endfunction()

google_cloud_cpp_set_enable_grpc_default()
```

it may work better *and* it seems easier to understand / debug. WDYT?

</blockquote></details>

Thanks I like this approach better.

I initially moved the helper function to `GoogleCloudCppCommon` but including that causes problems because of the Doxygen code. But then having a function seemed out of place, so I just left it as regular code.

**Review by devjgm — APPROVED** [review]
LGTM. I think this looks like a nice usability improvement, and I think it'll be easier for us to maintain going forward as we add more libraries. Thanks @remyabel !

I'd still like to await @coryan 's approval to merge in case he has any further comments.

**Review by coryan — APPROVED** [review]

**Pull request merged**
